### DISCUSSION TITLE
Sporadically failing with "Support for password authentication was removed" on GitHub

### DISCUSSION THREAD

**Discussion by Th3S4mur41** [semantic-release/semantic-release#3551] [Configuration]
### Current behavior

From time to time, the release job in the GitHub CI fails with "Support for password authentication was removed" error although the `GH_TOKEN` is set to a valid token.
Re-running the same job doesn't always solve the problem, but usually running a new job (e.g. through pushing an empty commit) does.

### Expected behavior

semantic-release should use the available token to identify to GIT

### `semantic-release` version

24.2.0

### CI environment

GitHub

### Plugins used

"@semantic-release/changelog": "^6.0.3",
"@semantic-release/exec": "^6.0.3",
"@semantic-release/git": "^10.0.1",
"semantic-release": "^24.2.0",
"semantic-release-major-tag": "^0.3.2"

### `semantic-release` configuration

```json
{
	"branches": ["main", "next", { "name": "beta", "prerelease": true }, { "name": "alpha", "prerelease": true }],
	"plugins": [
		[
			"@semantic-release/commit-analyzer",
			{
				"preset": "conventionalcommits",
				"releaseRules": [
					{
						"type": "build",
						"scope": "deps",
						"release": "patch"
					}
				]
			}
		],
		[
			"@semantic-release/release-notes-generator",
			{
				"preset": "conventionalcommits",
				"presetConfig": {
					"types": [
						{
							"type": "feat",
							"section": "Features"
						},
						{
							"type": "fix",
							"section": "Bug Fixes"
						},
						{
							"type": "build",
							"section": "Dependencies and Other Build Updates",
							"hidden": false
						}
					]
				}
			}
		],
		"@semantic-release/changelog",
		"@semantic-release/npm",
		[
			"@semantic-release/git",
			{
				"assets": ["CHANGELOG.md", "package.json", "package-lock.json"]
			}
		],
		"@semantic-release/github",
		[
			"@semantic-release/exec",
			{
				"publishCmd": "echo \"version=${nextRelease.version}\" >> \"$GITHUB_OUTPUT\""
			}
		],
		"semantic-release-major-tag"
	]
}
```

### CI logs

```bash
Run npm run release
  npm run release
  shell: sh -e {0}
  env:
    NPM_CONFIG_USERCONFIG: /__w/_temp/.npmrc
    NODE_AUTH_TOKEN: ***
    GITHUB_TOKEN: ***
    GH_TOKEN: ***
    NPM_TOKEN: ***
    HUSKY: 0

> @scope/playground@1.5.121 release
> npx semantic-release

[12:00:14 AM] [semantic-release] › ℹ  Running semantic-release version 24.2.0
(node:723) [DEP0180] DeprecationWarning: fs.Stats constructor is deprecated.
(Use `node --trace-deprecation ...` to show where the warning was created)
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/changelog"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/github"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "verifyRelease" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/changelog"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/git"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/github"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/github"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/github"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/exec"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "success" from "semantic-release-major-tag"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/github"
[12:00:15 AM] [semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/exec"
[12:00:16 AM] [semantic-release] › ✘  An error occurred while running semantic-release: ExecaError: Command failed with exit code 128: git ls-remote --heads 'https://github.com/***/playground'

remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/***/playground/'
    at getFinalError (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/return/final-error.js:6:9)
    at makeError (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/return/result.js:86:16)
    at getAsyncResult (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:164:4)
    at handlePromise (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:148:17)
    at async getBranches (file:///__w/playground/playground/node_modules/semantic-release/lib/git.js:69:11)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/lib/branches/expand.js:6:23)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/lib/branches/index.js:14:26)
    at async run (file:///__w/playground/playground/node_modules/semantic-release/index.js:68:22)
    at async Module.default (file:///__w/playground/playground/node_modules/semantic-release/index.js:278:22)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/cli.js:55:5) {
  shortMessage: "Command failed with exit code 128: git ls-remote --heads 'https://github.com/***/playground'",
  command: 'git ls-remote --heads https://github.com/***/playground',
  escapedCommand: "git ls-remote --heads 'https://github.com/***/playground'",
  cwd: '/__w/playground/playground',
  durationMs: 357.803113,
  failed: true,
  timedOut: false,
  isCanceled: false,
  isTerminated: false,
  isMaxBuffer: false,
  exitCode: 128,
  stdout: '',
  stderr: 'remote: Support for password authentication was removed on August 13, 2021.\n' +
    'remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.\n' +
    "fatal: Authentication failed for 'https://github.com/***/playground/'",
  stdio: [
    undefined,
    '',
    'remote: Support for password authentication was removed on August 13, 2021.\n' +
      'remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.\n' +
      "fatal: Authentication failed for 'https://github.com/***/playground/'"
  ],
  pipedFrom: []
}
ExecaError: Command failed with exit code 128: git ls-remote --heads 'https://github.com/***/playground'

remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/***/playground/'
    at getFinalError (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/return/final-error.js:6:9)
    at makeError (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/return/result.js:86:16)
    at getAsyncResult (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:164:4)
    at handlePromise (file:///__w/playground/playground/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:148:17)
    at async getBranches (file:///__w/playground/playground/node_modules/semantic-release/lib/git.js:69:11)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/lib/branches/expand.js:6:23)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/lib/branches/index.js:14:26)
    at async run (file:///__w/playground/playground/node_modules/semantic-release/index.js:68:22)
    at async Module.default (file:///__w/playground/playground/node_modules/semantic-release/index.js:278:22)
    at async default (file:///__w/playground/playground/node_modules/semantic-release/cli.js:55:5) {
  shortMessage: "Command failed with exit code 128: git ls-remote --heads 'https://github.com/***/playground'",
  command: 'git ls-remote --heads https://github.com/***/playground',
  escapedCommand: "git ls-remote --heads 'https://github.com/***/playground'",
  cwd: '/__w/playground/playground',
  durationMs: 357.803113,
  failed: true,
  timedOut: false,
  isCanceled: false,
  isTerminated: false,
  isMaxBuffer: false,
  exitCode: 128,
  stdout: '',
  stderr: 'remote: Support for password authentication was removed on August 13, 2021.\n' +
    'remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.\n' +
    "fatal: Authentication failed for 'https://github.com/***/playground/'",
  stdio: [
    undefined,
    '',
    'remote: Support for password authentication was removed on August 13, 2021.\n' +
      'remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.\n' +
      "fatal: Authentication failed for 'https://github.com/***/playground/'"
  ],
  pipedFrom: []
}
Error: Process completed with exit code 1.
```

**Comment by travi** [comment]
you likely need to [disable credential persistence](https://semantic-release.gitbook.io/semantic-release/recipes/ci-configurations/github-actions#pushing-package.json-changes-to-your-repository)

also, consider [whether you truly need to make a commit](https://semantic-release.gitbook.io/semantic-release/support/faq#why-is-the-package.jsons-version-not-updated-in-my-repository) during your release.

**Comment by Th3S4mur41** [comment]
Thanks @travi 

I double checked to be sure, but credential persistence was already set to `false` in the affected repositories.
Could the `GITHUB_TOKEN` still interfere even when setting `GH_TOKEN`?
So maybe setting both variable to the same value would fix this?

```yaml
      - name: Release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          GH_TOKEN: ${{ secrets.GH_TOKEN || secrets.GITHUB_TOKEN }} # Needed to push to GitHub
        run: |
          npm run release
```

**Comment by robfraz** [comment]
Hi, I've suffered a lot of sporadic failures with semantic-release, all of which are very similar to the one you posted.  I found turning on the debug logging (with `--debug`) very helpful in diagnosing what was actually happing, as the info-level logs were extremely misleading.

For me, it turned out that ALL of the sporadic failures were caused by two or more pull-requests getting merged in a short timeframe (e.g. within a couple of minutes of each other):

- The first merge triggers semantic release CI job (we'll call this `job A`) to start running.
- Before `job A` gets very far, a second merge occurs that starts up semantic release `job B`.
- `job A` is now running against a commit that is behind the tip of the release branch.
- `job A` now reaches the point where it's actually running semantic-release code, and one of the very first things it does is to verify its authentication as you can see [here](https://github.com/semantic-release/semantic-release/blob/v25.0.2/index.js#L86-L102).
- Semantic release verifies that it has authentication by using `git push --dry-run --no-verify` (see [here](https://github.com/semantic-release/semantic-release/blob/v25.0.2/lib/git.js#L196-L212)), which fails on `job A` not because of an actual authentication problem, but because the commit is no-longer the latest commit found on the branch (you cannot push to a remote branch if the commit you are pushing is behind).  Note that a failure here only gets sent to the debug logs.  For `job B`, this should succeed, because it is using the latest commit.
- If authentication fails, `job A` then [checks](https://github.com/semantic-release/semantic-release/blob/v25.0.2/index.js#L90) to see if see if the local branch is up-to-date with the remote one, by using `isBranchUpToDate()`, defined [here](https://github.com/semantic-release/semantic-release/blob/v25.0.2/lib/git.js#L292-L297), which uses `git ls-remote --heads`, as seen in your error logs. However, for reasons I don't yet understand, this fails with the misleading authentication errors, even when you have valid auth. For me, the failures I get seem to imply that `git` has fallen back to a different authentication method that I don't have configured, but like I say, I still don't quite understand this part of the problem.
- `job B` succeeds, and a release is made.
- `job A` will never succeed on a retry, because it is always using a commit that is behind.  And as you have observed, if you merge a new commit to the repo, semantic release magically starts succeeding again because it is working with a commit that is no-longer behind the tip of the branch, and all the authentication issues it was previously reporting in `job A` were all lies.

And because `job B` succeeded (using the later commit), it doesn't actually matter that `job A` failed, but it's still a nuisance regardless, and causes me no-end of problems with people falsely reporting authentication failures in our build system.  

**In summary:** it seems to me that the semantic release codebase knows that using `git push --dry-run --no-verify` can fail if the branch is not up-to-date, and tries to detect if that is the case, but something then goes wrong whilst checking if the branch is up-to-date. This results in misleading authentication errors, when the only problem is that it's working with a commit that's no-longer at the tip of the release branch.
